### PR TITLE
Coderefs endpoints

### PR DESCRIPTION
**Pull request by bttf**
### Features and Changes
- Adds new data model `FeatureCodeRefs`
	- Meant to pair a feature flag with multiple code references for a given repo/branch
	- Unique compound index on `organization`, `repo`, `branch`, `feature`
- Adds two new endpoints to REST API
	- `GET /api/v1/feature-keys`
		- For retrieving a flat list of all feature keys for an organization.
		- Response is an array of strings.
	- `POST /api/v1/code-refs`
		- For submitting code references generated by `gb-find-code-refs`.
		- Accepts query parameters for required metadata:
			- `repo`
			- `branch`
			- `platform` (optional)

### Testing
Tested using cURL requests

For the GET endpoint
`curl -H "Authorization: Bearer secret_admin_tgIWDS2bk1JRyhvkIMRqM9HfbFuoWr5owGoJWceBrHw" localhost:3100/api/v1/feature-keys`

For POST
`curl -XPOST -H "Authorization: Bearer secret_admin_tgIWDS2bk1JRyhvkIMRqM9HfbFuoWr5owGoJWceBrHw" -H "Content-Type: application/json" -d @coderefs_adnan_coderefs-endpoints.json "localhost:3100/api/v1/code-refs?repo=growthbook%2Fgrowthbook&branch=main"`

### Screenshots
FFs in GB
image

Fetching keys from `/feature-keys`
image

POSTing coderefs to `/code-refs`
image
 
Data stored in DB
image

### CODE REVIEW

File: packages/back-end/src/models/FeatureCodeRefs.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+};
+
+export const getAllCodeRefsForFeature = async ({
+  feature,
+  organization,
+}: {
+  feature: string;
+  organization: OrganizationInterface;
+}): Promise<FeatureCodeRefsInterface[]> => {
+  if (!feature) throw new Error("Missing required parameters");
```
Comment (mknowlton89) [inline packages/back-end/src/models/FeatureCodeRefs.ts:130]: I'm not sure this check will ever run - do we need it?

Comment (mknowlton89) [inline packages/back-end/src/models/FeatureCodeRefs.ts:130]: I'm not sure this check will ever run - do we need it?

File: packages/back-end/src/models/FeatureCodeRefs.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  organization,
+}: {
+  feature: string;
+  repo: string;
+  branch: string;
+  platform?: "github" | "gitlab" | "bitbucket";
+  codeRefs: FeatureCodeRefsInterface["refs"];
+  organization: OrganizationInterface;
+}): Promise<FeatureCodeRefsInterface[]> => {
+  if (!feature || !repo || !branch || !codeRefs) {
```
Comment (mknowlton89) [inline packages/back-end/src/models/FeatureCodeRefs.ts:58]: Is this check necessary?

File: packages/back-end/src/models/FeatureCodeRefs.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  features,
+  organization,
+}: {
+  repo: string;
+  branch: string;
+  platform?: "github" | "gitlab" | "bitbucket";
+  features: string[];
+  organization: OrganizationInterface;
+}): Promise<FeatureCodeRefsInterface[]> => {
+  if (!repo || !branch) {
```
Comment (mknowlton89) [inline packages/back-end/src/models/FeatureCodeRefs.ts:105]: Same here - is this needed?

**Review by mknowlton89 — APPROVED** [review]

**Review by mknowlton89 — APPROVED** [review]

File: packages/back-end/src/api/openapi/payload-schemas/PostCodeRefsPayload.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      properties:
+        filePath:
+          type: string
+        startingLineNumber:
+          type: integer
+        lines:
+          type: string
+        flagKey:
+          type: string
+        kontentHash:
```
Comment (bttf) [inline packages/back-end/src/api/openapi/payload-schemas/PostCodeRefsPayload.yaml:30]: TODO Fix typo

File: packages/back-end/src/models/FeatureCodeRefs.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+import { FeatureCodeRefsInterface } from "../../types/code-refs";
+import { OrganizationInterface } from "../../types/organization";
+
+const featureCodeRefsSchema = new mongoose.Schema({
+  organization: String,
+  dateUpdated: Date,
+  feature: String,
+  repo: String,
+  branch: String,
+  platform: {
```
Comment (bttf) [inline packages/back-end/src/models/FeatureCodeRefs.ts:12]: TODO Remove from model; this is captured in org settings instead.

File: packages/back-end/src/models/FeatureCodeRefs.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+      platform,
+      dateUpdated: new Date(),
+    },
+    { upsert: true }
+  );
+
+  return await FeatureCodeRefsModel.find({
+    feature,
+    repo,
+    branch,
```
Comment (bttf) [inline packages/back-end/src/models/FeatureCodeRefs.ts:76]: TODO include org here

**Comment by bttf** [comment]
TODO Add existing getting started steps to docs

**Pull request merged**